### ISSUE TITLE
Media library needs attractive masonry layout

### ISSUE THREAD

**Issue by jauyong** [GoogleForCreators/web-stories-wp#248]
As a user I want to view the files in my media library in a aesthetically pleasing way

**Acceptance Criteria**
1. Files in the media library utilize a masonry layout
3. Needs to work with infinite scroll (without reordering)

**[Feature Brief](https://docs.google.com/document/d/1aZH_TI_be2t5Ywyqx4jcPTPdhFGEt7xh-uLzBp2AV4A/edit#heading=h.xqcejitqohdk)**
**[Design](https://www.figma.com/file/Cd9iacVknHK62RO4pRwb68/Stories-Editor?node-id=2731%3A1757)**

**Comment by spacedmonkey** [comment]
In my research, [packery](https://packery.metafizzy.co/) seem pretty interesting.

**Comment by swissspidy** [comment]
That one might not work due to licensing

**Comment by spacedmonkey** [comment]
Another option here would be using [brick.js](https://github.com/callmecavs/bricks.js)

**Comment by barklund** [comment]
We have a very simple use-case for masonry here - only two columns and all elements will be single-column width. Plus we need accessibility in terms of keyboard navigation between elements. I can't actually find any library with the latter, and the former is extremely simple to do.

We basically just need to remember the height of each column and put any new element in the shortest column. That's very very basic.

And then we do accessibility by having up/down move inside the column and left/right move between the columns to the element, that is right across from the middle of the currently focused element. I think I can write that pretty succinctly without any external library in rather few lines.

I recommend we implement this custom.

**Comment by swissspidy** [comment]
After checking out that bricks.js library and seeing how simple the implementation is (despite it having more features than we need) I tend to agree with this 👍

**Comment by wassgha** [comment]
@barklund can we verify that that's a standard way to do accessibility in this interaction? Otherwise a simpler way to do it would be to add a `tabindex` equal to the real index of the media element and have the browser handle tabbing.

**Comment by barklund** [comment]
> @barklund can we verify that that's a standard way to do accessibility in this interaction? Otherwise a simpler way to do it would be to add a `tabindex` equal to the real index of the media element and have the browser handle tabbing.

@wassgha we can't do tabbing, only arrows. As the media library has infinite scroll, you'd effectively block all additional tab targets on the page from ever being reached - you'd be stuck tabbing "forever".

Accessibility guidelines on more complex apps also recommend tabbing between navigational sections, but arrow navigation within those sections (this include Google's own guidelines that [@dvoytenko has summarized](https://docs.google.com/document/d/1jR6c0eGNfoUXJg04v_isMfPnvRc2SD8zfevCIPqAjF4/edit?usp=sharing)).

Besides, tabbing would also quickly go out of sync with spatial location if odd elements were significantly shorter or larger than even elements.

Arrow navigation is definitely the way to go here.

**Comment by spacedmonkey** [comment]
> After checking out that bricks.js library and seeing how simple the implementation is (despite it having more features than we need) I tend to agree with this 👍

Do this mean you pro or against using the library?

**Comment by swissspidy** [comment]
@spacedmonkey Ah, my bad, it wasn't really clear. I'd go for a custom implementation.

**Comment by wassgha** [comment]
> > @barklund can we verify that that's a standard way to do accessibility in this interaction? Otherwise a simpler way to do it would be to add a `tabindex` equal to the real index of the media element and have the browser handle tabbing.
> 
> @wassgha we can't do tabbing, only arrows. As the media library has infinite scroll, you'd effectively block all additional tab targets on the page from ever being reached - you'd be stuck tabbing "forever".
> 
> Accessibility guidelines on more complex apps also recommend tabbing between navigational sections, but arrow navigation within those sections (this include Google's own guidelines that [@dvoytenko has summarized](https://docs.google.com/document/d/1jR6c0eGNfoUXJg04v_isMfPnvRc2SD8zfevCIPqAjF4/edit?usp=sharing)).
> 
> Besides, tabbing would also quickly go out of sync with spatial location if odd elements were significantly shorter or larger than even elements.
> 
> Arrow navigation is definitely the way to go here.

In that case I agree with a custom implementation 👍

**Comment by spacedmonkey** [comment]
Okay then, what we have there should work, then we just been work workout tab ordering.

**Comment by spacedmonkey** [comment]
@swissspidy @pbakaus It seems like the group is learning towards a custom solution here. The current solution is a very simple ( and a little hacky ). The media element are just broken into two columns. 
<img width="1884" alt="Screenshot 2020-02-25 at 10 00 27" src="https://user-images.githubusercontent.com/237508/75236451-b5433880-57b5-11ea-8408-423d4c2c83d4.png">

I wanted to implement a masonry library to maintain the semantic order for tabbing. 
But as @barklund has noted, 
> Arrow navigation is definitely the way to go here.

So maybe we should just stick with what we have here and just implement the new keyboard shortcut handler to navigate between the elements. This will be blocked by #26 as UX is required for focus / hover before we can continue with keyboard navigation.

**Comment by dvoytenko** [comment]
I agree with custom path: we should have all the info we need to decide on the prev/next item in the layout based on up/down keys. 

The important part of this component will be managing visibility of items in the layout to reduce amount of loaded resources. This is especially true for cases where we don't have video placeholders and have to load video elements in their place.

> This will be blocked by #26 as UX is required for focus / hover before we can continue with keyboard navigation.

I don't see why keyboard navigation should be blocked by this, however. Focus state and keyboard handlers is all we need for navigation itself. No?

**Comment by jauyong** [comment]
@jauyong to add brief to ticket description

**Comment by spacedmonkey** [comment]
A couple of libraries I have found in my research. 

http://callmecavs.com/bricks.js/ 
http://macyjs.com/
https://bvaughn.github.io/react-virtualized/#/components/Masonry

Also a CSS only solution.

**Comment by spacedmonkey** [comment]
Doing a little research into virtual lists, this could be a nice solution. But it way out of scope of this issue. I have create a new issue this at https://github.com/google/web-stories-wp/issues/1302 . This ticket can be sized etc separately.

**Comment by jauyong** [comment]
To redue scope of this ticket, we are removing accessibility work:
7. Implement keyboard navigation. Keyboard should trigger infinite scroll loading for next page.  

@jauyong to create a new ticket for this AC.

**Comment by jauyong** [comment]
Created #1305 for accessibility features.
Closing this as masonry AC has been added to 1302.
